### ISSUE TITLE
Why your previous developer was terrible

### ISSUE THREAD

**Issue by MasterAtNone91** [iinf362-cjv/Assignments#1]
Post your takeaway from the article mentioned in class. It can also be found here https://medium.com/things-developers-care-about/why-your-previous-developer-was-terrible-506a06ae35ea

**Comment by cjfrancis13** [comment]
Shamoon Siddigui makes a great argument when he pinpoints to his idea of "the curse of the present", which is the more developed understanding or method of approach a developer would use given that a framework is already established. Most coders and developers tend to get shamed for setting the building blocks incorrectly, which in my opinion, is an unfair criticism. When initiating a project for the first time, it has to be understood that there will be a learning curve to it and many cases of trial-and-error. There is a level of asymmetric knowledge at play here, especially when talking about the "scapegoat effect" the new developer instills towards the old developer. The new developer is clearly placed at an advantage knowing that whatever approach the old developer implemented wasn't a best practice.

**Comment by LorennaFrancelino** [comment]
I really loved this article! That's really what happens when a developer leaves a project, actually this happened with me. I only had one experience as a developer, I worked in a TV company in my country, Brazil, and, when I started working there, I had no experience at all. It was really my one and only experience! I had to learn a new programming language by myself, and they gave me a brand new project to work on. Basically, I had to join a new device to an existing software, that would be controlled by this device. Because everything was new for me, and I had to learn things mostly by myself, it took me some time to get everything together, but once I got it, I could really do a good job. At least I was happy what I was doing, as well as my boss was.

The problem is that I had to leave my job before finishing the project - I had the chance to come to the US to study abroad - and a friend from my university in Brazil started working there even before I traveled.  Well, he didn't say anything to me, but I heard from other friends that he was changing everything on the project, improving the code, removing unnecessary code, and so on. Actually, I got happy that he was doing well on the job, because I didn't want to leave the project without support. However, it was really easy for him to make changes, and improve what I had done, but what he did not know was that I built something where there was nothing before. When he got the job, he had material to base his steps on, unfortunately I didn't have.

In conclusion, as it was said in the article, I think that's why we can't never judge and/or blame the previous developer from his/her mistakes. It happens with everyone, we all can be in both positions one day... You never know!

**Comment by CSAlexander21** [comment]
I found this article to be very eye opening. When initially reading, I thought that first developer in the "short" story had no idea what he was doing and lacked knowledge of the project. As the newly hired developer made more efficient changes, I realized that she was able to made those suggestions solely because of the previous developer's groundwork. The "hacked job" committed by the first developer  could have been decent work considering the difficult circumstances of initiating a new project. It goes to show that in an IT project there are always ways for systems to be improved. In order to achieve the best end result it takes knowledge and observations from multiple supporters. With that being said,  you should not badmouth a previous developer because they are basically the pioneers of the project. When the new developer passes blame it is unfair because their work was used as the a foundation to complete the project.

**Comment by kevingsp** [comment]
I believe this is true in all aspects of starting a new project, whether it has to do with programming, designing, constructing or some other type of creating. The new person on the job will always critique the previous one, for reasons that Shamoon Siddiqui mentioned. Although I have no hands on experience with developing new software, libraries, or databases, I have taken over some unfinished projects and always questioned myself as to how the previous person was handling it. Everyone has their own way of processing tasks and requests so we all will handle situations differently unless there is some sort of strict protocol that has to be followed.

New developers fresh on the job will be coming on with fresh, outside knowledge and with all these new frameworks and languages that seem to be coming out every week. What I got from the article is that we, as students, should try to stay on top of all this new information that comes out. We need to be up to date with all the languages in order for us to try and prevent the next person from bad mouthing our work.

**Comment by MJAmk** [comment]
I believe Shamoon Siddiqui made a great point. If the boss gave the previous developer a deadline that the person needed to meet then they would do their best to meet the deadline. If the deadline made them need to cut corners and do something in a strange way then it is not necessarily the fault of the developer. Missing a deadline is something that many people would not like to be in the position of. The previous developer could have thought it more beneficial to make something as a quick temporary solution than an actual solution.

I agree with not using the previous developer as an excuse for your own laziness or inability to do something. There can be many mistakes, but it is best to not blame others every time. An example of something that happens often in the world is selling a "finished" video game broken. A lot of times video games get released with bugs because the developers of the game are given a deadline that they need to meet. Just because something the previous developer made is not the best solution does not mean the previous developer wanted to make it that way. There are external factors that can affect everyone, so current developers should not always blame the old developer.

**Comment by aburke2** [comment]
I agree with Shamoon Siddiqui on a number of his points. I can especially relate to the "Pass The Blame" portion of the article. Although I don't have experience as a developer or software creation, I have experience working with project managers as well as developers. I felt that when there were delays in a project regarding software changes, one developer would blame another and vice versa. By passing the blame onto others, it can actually hurt the project from a deadline point of view--which may create some tension or frustration between all parties involved.

Siddiqui also mentioned that new developers should refrain from making huge changes early on, as it's only a method of justification used to impress your new boss. I agree with this, and I also think that it's important to really understand old processes before making huge, seemingly pivotal, changes. I think that the new developer's knowledge could be showcase in other ways, perhaps later on or in small intervals. Towards the end of the "rant," Siddiqui mentions that the stakeholders should all be aware of what changes are being made and next steps. I think that is extremely important not only from the task at hand, but also branding yourself as a reliable professional in the field.

**Comment by mapleljn** [comment]
I like this article, Shaman Siddiqui made a great point about the curse of the present. In my opinion, most developers will choose the best tool to do their work, not all tools that new developers use are available for previous developers, and it’s not easy to change to a new language such as Node.js when an app was already built with Rails.
I agree that new developers should not blame the previous developer, everyone makes mistakes. Previous developer might already tried their best, so it’s better for new developer do their best to prove their value instead of blame the previous one.

**Comment by 9one9** [comment]
Siddiqui has definitely brought some sense to the table on programming. There is a blame game in the industry and it's not necessary to be a victim to perceive. With the tools used to program every day evolving, the approach to the programming and the the people in front the monitor can only adapt so much. There is an advantage to the programmer on salary, he/she can continue writing code to their philosophy. I agree there is no perfect programmer, however the blame shifting stops on the better programmer. 

Making good code depends on hard work and toil; doing those big changes in the beginning makes a statement ONLY when functional set in stone. The strongest and best changes REDUCE the code and the burden on any programmer. An efficient program is the goal of any developer.  With that being said, the best programmer will make code rigorous and would stand through the test of time.

**Comment by jblieden94** [comment]
Siddiqui makes some great points when it comes to why your previous developer was not as good as you thought. Whether is was the, "curse of the present" or some other unforeseen problem, the previous developer was to blame. The curse of the  present was a great was to show that it might not be entirely the previous developer's fault. This example could be implemented in many other things, concepts and ways to code could have drastically become easier with time but the previous developer already finished the project and didn't want to do it all over again. The way this new developer passes the blame to his/her predecessor is easy. Everyone does it in every job, its not harder when its coding its easier. The curse of the present makes this easy because new and more efficient ways to code are becoming more and more available. Overall Siddiqui's really opened up how much developers actually have to work because they might be fixing mistakes or just updating what the previous developer has done.

**Comment by Merln4052** [comment]
I agree with Siddiqui on his point about how we should not play the blame game with new developer vs old developer. Its unfair because they probably did make the best decision that they could at the time. As well as the technology that the new developer is using or trying to implement may not have been around or available to the old developer. With coding only getting easier it only makes the balame game easier. It really is the curse of the present in the sense of the newer/better ways of developing things has only made it easier for the developer to say "it's not my fault that they built your product using this when they should have used that." As these decisions are already made with someone else's time and effort.

**Comment by ellaorzhu** [comment]
This article is interesting to read. I totally agree with the author’s point of view. There are a few factors that blame is not supposed to happen. Firstly, Knowledge is upgraded dramatically. Old tool is replaced by new tool. It is reasonable that an application created by previous developer early on need to be changed in the present. Thus we definitely can not blame previous developer for using tool which we seem to be out-of-date and inappropriate for current technology. Secondly, a new developer who comes to the position should respect previous developer and his or her work. A new developer might do not realize that how much time a previous developer or team devoted to project and how hard they work. A new developer is not able to deny contributions which previous developers did early on. As a new developer, he or she is responsible for doing well on tasks assigned by boss, not blame how terrible previous developer is. Indeed, new developer who blames previous developer is an easy way to make excuses for his or her incompetence or laziness.
      How to end it？it is hard to say. Because the action that we decide to do is always harder than what we say. New employees tend to perform well and make a good impression in front of boss. Therefore, it is likely to blame previous developer when working on old version in order to show that they are capable.

**Comment by ellaorzhu** [comment]
This article is interesting to read. I totally agree with the author’s point of view. There are a few factors that blame is not supposed to happen. Firstly, Knowledge is upgraded dramatically. Old tool is replaced by new tool. It is reasonable that an application created by previous developer early on need to be changed in the present. Thus we definitely can not blame previous developer for using tool which we seem to be out-of-date and inappropriate for current technology. Secondly, a new developer who comes to the position should respect previous developer and his or her work. A new developer might do not realize that how much time a previous developer or team devoted to project and how hard they work. A new developer is not able to deny contributions which previous developers did early on. As a new developer, he or she is responsible for doing well on tasks assigned by boss, not blame how terrible previous developer is. Indeed, new developer who blames previous developer is an easy way to make excuses for his or her incompetence or laziness.
       How to end it？it is hard to say. Because the action that we decide to do is always harder than what we say. New employees tend to perform well and make a good impression in front of boss. Therefore, it is likely to blame previous developer when working on old version in order to show that they are capable.

**Comment by ellaorzhu** [comment]
This article is interesting to read. I totally agree with the author’s point of view. There are a few factors that blame is not supposed to happen. Firstly, Knowledge is upgraded dramatically. Old tool is replaced by new tool. It is reasonable that an application created by previous developer early on need to be changed in the present. Thus we definitely can not blame previous developer for using tool which we seem to be out-of-date and inappropriate for current technology. Secondly, a new developer who comes to the position should respect previous developer and his or her work. A new developer might do not realize that how much time a previous developer or team devoted to project and how hard they work. A new developer is not able to deny contributions which previous developers did early on. As a new developer, he or she is responsible for doing well on tasks assigned by boss, not blame how terrible previous developer is. Indeed, new developer who blames previous developer is an easy way to make excuses for his or her incompetence or laziness.
How to end it？it is hard to say. Because the action that we decide to do is always harder than what we say. New employees tend to perform well and make a good impression in front of boss. Therefore, it is likely to blame previous developer when working on old version in order to show that they are capable.

**Comment by sadex3** [comment]
We are very pleased once we have found a solution to our initial problem. This article was very eye opening and realistic. In the beginning of the article,  the author was describing the moment after an employer realizes, what of a great job they did with choosing their new developer. The employer fails to realize that the new developer is an outsider looking in. The curse of the present, was very much something I can relate to. It pinpoints the new hires fresh knowledge, but attitude towards not wanting to be blamed for the issues that occurred because of the past developer. While reading, it finally made sense, the article stated, "When the previous developer (or team) had to develop it, they had to deal with a LOT of unknowns. They had to make many decisions under a cloak of opacity." Your knowledge is only tested and good enough to fix the errors you see, but not the one's you create. In this situation, I believe it is good to do pairing programming. Pair programming is when two developers work together on one workstation. "The “driver” is the one writing the code, and the other person (most often called the “navigator”) reviews the code as it’s written." The roles can change at anytime. I believe this method is great to use. It allows one person to focus on the current line of code, while the other person can think about the overall design. Having a second set of eyes more than pays for itself in fewer bugs down the road.

**Comment by cjs166** [comment]
This article to me was a very interesting read and I agree with the sentiments of Siddiqui all the way.  He highlighted a problem in the world of developers that i never really thought to think of.  In the era we live in everything is advancing so quickly when it come to the world of technology, software and coding.  There are always new ways to simplify our work and our daily lives popping up all year round.  So it is almost inevitable to look at a programming project started a few years and not look at it like it has mistakes and could of been made more simpler.  Theres been new programs and dramatic software changes made just in the last 3-5 years.  This just shows you should always be trying to learn more about some of the new innovations coming about and revising your code when ever possible.  You might have the next guy thats just came out of college with new and fresh information bad mouthing you so he can take your position which is totally wrong.

**Comment by gregbedik** [comment]
Shamoon Siddiqui's article touched on some key points in todays technological advanced society.  "The Curse of the Present" is not just something he made up, but a very true practice that happens in every day life.  One of the biggest challenges we face is no matter what you do to stay onto of the technology and the current trend, it always changes.  There will be something newer and better out tomorrow.  This causes a challenge for developers because they have a deadline and need to come up with a product, but when they first started to now, there are new technologies out that, will make the past practices used, questionable.  Also, as mentioned in the article, there are a lot of unknowns when you are the creator of something.  Therefore, you use what you know to be best and current for the time you are creating the product.
Another point he made that is relevant in almost everything we do today, is the idea of passing the blame.  When you are tasked in creating something or updating something and the product is bad, you will tend to bad mouth the initial creator and question everything they did.  Although it might seem easier to do this, it is more beneficial to you and the outcome of the project to just identify the problems and just fix them.  Rather then passing the blame and delaying your process to put down someone who isn't even there, identifying the problems in a positive way and understanding when this was created there might not of been proper technologies for it, will allow for  a better outcome.

**Comment by NickEngert** [comment]
This article touches on some very key points in today’s society. Today, technology is ever-growing, and always improving to make things easier from past products and services. The term “The Curse of the present” is something he touched upon, basically saying that often when previous developers design a system, it seems they did a horrible job when the new developer goes to improve and fix it. But, on the contrary they did the best of their ability (you hope) under the restraints and time they had, with the knowledge and resources they had. There are always unknowns in developing something, so they often have to make their best judgement and hope these ‘unknowns’ are something they accounted for. Lastly, they talk about ‘passing the blame’, or what happens when a new developer comes in and badmouths the previous developer’s choices without taking into account the technology available to them when they originally developed it. He ends that we as developers or future ‘new’ developers to instead of badmouthing the previous, just improve and fix the problems. This will help everyone in the long run.

**Comment by VeraYang** [comment]
I agree with Shamoon Siddiqui’s point about the curse of the present. I think previous developers should not be blamed if they tried their best to finish the job before the deadline. It’s important for us to realize that the information and technology we can access to are much different from the past. Moreover, since we are “standing upon the shoulders of Giants”, it’s pointless to be badmouth. And as for new developers, what we need to focus on is to develop the system with better methods in order to meet the requirement of the boss, even the times, instead of blaming previous developers or using them as an excuse for not finishing the work.

**Comment by sguthrie113** [comment]
This article does a great job explaining why we should not downplay others efforts. We must view others' work with the mindset that they used the most efficient method at the time. Our field is constantly changing and updating. Many codes and programs that seem like the most sensible for us now will be outdated within a couple months. As developers we should be less judgmental and more understanding, looking out for each other as much as possible.

**Comment by waveywednesday** [comment]
The moral of this article can be put into numerous situations and applied in all different aspects of life. As Siddiqui so very well explained, it is much easier for us to claim a scapegoat and put the blame on others than it is for us to accept the wrongs and correct them. If you are taking over something that a previous person was handling it is always easier to make yourself look like the better option in the matter, but that is not the way to go about it. When it comes to a craft that someone takes very seriously and spent numerous amount of hours perfecting, talking down on them diminishes their accreditation and also can create serious feuding between developers. All in all, Siddiqui makes a very valid point that is true in all aspects of life that can and should be applied on a daily basis.

**Comment by japodolla** [comment]
“Why your previous developer was terrible and why your current one seems so amazing” was a very interesting read. He brought attention to common behaviors of previous and current developers. For example, passing the blame. IT’s extremely easy to fault someone who isn’t there to defend him/herself. Here is where the author hit the nail on the head, “When the boss asks “what’s the timeline on the [whatever],” it’s easy and convenient to say “well, normally it would only take 2 weeks, but since we’re dealing with an older version of [some library], it’ll probably take a month.’” I’m guilty of such behavior. He made several valid points. It is important that we don’t throw others under the bus in order to be the hero of the moment.

**Comment by nmartucci** [comment]
Shamoon Siddiqui Did a great Job explaining the moral trials faced by new and upcoming developers. Bashing a previous developer for the choices they made in their system is far from okay. As Siddiqui stated that "I think it’s a bad practice to simply blame the previous developer or team for something. Give them the benefit of the doubt and assume that they made the best decisions they could under the constraints they had...Be a hero in the long run by being a solid team player that makes good judgement calls whenever you can. Don’t be the short-term hero that throws people under the bus." I fully agree with him. It would be bad practices if one were to bash someone for the work they did to the best of their ability with restrictions or constraints present. My reasoning is that eventually you'll be at the other end of that hypothetical stick in the future when better methods are uncovered as well as better developers come to light.

**Comment by kphil120** [comment]
After reading this article I believe that Shamoon Siddiqui rant was believable and eye opening. He justifies the fact that people are lazy and will do anything in there power to do less work. However that's not their fault, its the fact that it is human nature to be lazy. In the IT world the simplest way to ease a heavy workload as a developer would be to blame the developer before. However in this case the truth may be taken out of context. As a developer you should work with what you have and try to improve on the problems at hand. Yes it may be easier to blame the previous developer but as Shamoon Siddiqui stated it is bad practice to do that. All in all as developers we must look past the fault of the previous developers but look toward the future of how to become better then the previous developer.

**Comment by shaniceD** [comment]
I believe Shamoon Siddiqui made a great point. The "Pass The Blame" portion of the article is what i can relate to the most. Although most of the class already said this i would have to say i stronger agree with them, If there is a deadline to met then if is perfectly understandable that the developer tries their best to meet say deadline and if meeting that deadlines requires him to take a short cut or leave out a few complex things then it is not 100 percent the fault of the developer. The developer worked as best as he/she could to make the product in a timely fashion.

   But i can honestly say I’ve been on both sides of the story. I’ve been the developer who’ve disagreed with previous developers work, and I’ve also been the person who another had to work on projects i couldn’t finish or needed help with. And to be honest: When you’re a developer or someone who’s learning to develop, looking at an existing codebase, you can spot a lot of bad things. You’ll probably spot a lot of things that aren’t necessarily bad but you might think they are bad code, but it's simply different than how you would’ve solved it.

**Comment by jwoa** [comment]
There are a lot of strong points in this article. But, the primary one that sticks to me is the idea of not blaming the past developer for what's left because you never know what they had to go through to reach it. Sure, when the new person comes in the items may seem useless and unnecessary but maybe at the time that they were implemented those items were perfect. It always pays to be a great team player at the end instead of dragging somebody else down. People need to take into consideration the amount of thought and work that went into it, of course the developer wouldn't slack in purpose. But overall Shamoon Siddiqui made a lot of strong valid points.

**Comment by jasminekimber91** [comment]
In the article Shamoon Siddiqui mentions some bad issues that some new developers might meet when they come to a new position. Shamoon Siddiqui also expressed his point of view about why some newdevelopers may blame the old developer. One reason is that the new developer may want to pass blame to the old developer. I really agree that Shamoon explains the reasons about why new developers tend to pass blame to previous developer. Shamoon says because the previous developer is no longer here to defend himself so that he is super easy to be blamed by new developers. I believe what Shamoon mentions in the article really happens in real life. I think it's so unfair to the previous developer and it's so mean for some new developers to do this.
     There is one sentence in the article really impressed me, Shamoon says "Be a hero in the long run by being a solid team player that makes good judgement calls whenever you can. Don’t be the short-term hero that throws people under the bus" Shamoon shares his opinion of some moral things  by using this sentence,  I think this sentence means if you want to be a good developer, at first you have to think yourself as a team, you have learn how to unite and honest. Don't make excuse for yourself and make bad judgement on other people just because you want to be a short-term hero or pass the blame.

**Comment by ckelly721** [comment]
Shamoon Siddiqui expresses a valid argument about his idea of “the curse of the present”. When individuals usually start a new job, they sometimes feel as if they have to pick up the slack of the previous employee. With this in mind, badmouthing is often expected, either from the new employee or other employees. In the case of developers, new technology is constantly being created, so it is quite easy for the new developer to feel as if the previous developer made poor decisions in choosing certain programs. Siddiqui rants about this because it is a recurring situation within his career field as well as within others. His solution, which I completely agree with, is to not throw previous employees under the bus. Instead, be a solid team player and make good judgments when it calls for it.

**Comment by Chibuikem** [comment]
I read the article, and I found it very interesting. I haven't been in the position of being a developer, working on a platform that has already been developed, or being the first one to create something. I can imagine how it would feel because I think it is a way of making things easier as stated in the article. Also, I feel that the enthusiasm a new developer would have coming into the team and making all these BIG changes to show he/she knows what he/she is doing, is the same level the previous developer had when creating the platform so I agree with the article when she says "I think it’s a bad practice to simply blame the previous developer or team for something. Give them the benefit of the doubt and assume that they made the best decisions they could under the constraints they had." 

Like the previous assignment Death to Bullshit, this article has given me a new mindset. What to avoiding doing and how to avoid acting when I come into a new situation such as a developer for a team or project, this article can also be applied to other things in life other than "Why Your Previous Developer was terrible" although your previous developer might have been terrible.

**Comment by AmandaTReis** [comment]
The author is absolutely right and I can`t agree more. Unfortunately, not just developers do this. In my opinion, we can`t judge the previous developer just because it is convenient on that moment to do so. Mostly the new developer is not familiar with the technology that is being used and concluded that the technology is not the best to solve the problem. In addition, I think that the new developer should put himself in the other developer position before judge. Thus, analyze the previous situation and what may have influenced the choice of the previous developer.
